util/eventbus/eventbustest/eventbustest.go

```diff
@@ -201,3 +201,40 @@ func eventFilter(f any) filter {
 		return fixup(fv.Call([]reflect.Value{args[0].Elem()}))
 	}).Interface().(filter)
 }
+
+// Injector holds a map with [eventbus.Publisher], tied to an [eventbus.Client]
+// for testing purposes.
+type Injector struct {
+	client     *eventbus.Client
+	publishers map[any]any
+}
+
+// NewInjector constructs an [Injector] that can be used to inject events into
+// the the stream of events used by code under test. After construction the
+// caller may use [Inject] to insert events into the bus.
+func NewInjector(t *testing.T, b *eventbus.Bus) *Injector {
+	inj := &Injector{
+		client:     b.Client(t.Name()),
+		publishers: make(map[any]any),
+	}
+	t.Cleanup(inj.client.Close)
+
+	return inj
+}
+
+// Inject inserts events of T onto an [eventbus.Bus]. If an [eventbus.Publisher]
+// for the type does not exist, it will be initialized lazily.
+func Inject[T any](inj *Injector, event T) {
+	eventType := reflect.TypeOf(event)
+
+	if _, ok := inj.publishers[eventType]; !ok {
+		inj.publishers[eventType] = eventbus.Publish[T](inj.client)
+	}
+
+	pub, ok := inj.publishers[eventType].(*eventbus.Publisher[T])
+	if !ok {
+		panic("could not find publisher")
+	}
```

Comment:
I think we can simplify this:
```suggestion
    pub, ok := inj.publishers[eventType]
	if !ok {
		pub = eventbus.Publish[T](inj.client)
		inj.publishers[eventType] = pub
	}
```

Reply:
Kind of, I need to cast still as `pub` turns into an `any` in this example. I can fix it by doing something like:

```
	castPub := pub.(*eventbus.Publisher[T])
	castPub.Publish(event)
```

after your change which seems reasonable.